### PR TITLE
Added support for complex objects in MDLC and NDLC using ObjectHandle

### PR DESCRIPTION
**Pull request by snakefoot**
Resolves https://github.com/NLog/NLog.Extensions.Logging/issues/237

### CODE REVIEW

File: src/NLog/MappedDiagnosticsLogicalContext.cs [hunk 1]
```diff
@@ -129,9 +132,17 @@ public static string Get(string item, IFormatProvider formatProvider)
...
-            object value;
-            GetLogicalThreadDictionary().TryGetValue(item, out value);
+            if (!GetLogicalThreadDictionary().TryGetValue(item, out var value))
+                return null;
+
+#if NET4_6 || NETSTANDARD
             return value;
+#else
+            if (value is System.Runtime.Remoting.ObjectHandle objectHandle)
+                return objectHandle.Unwrap();
```
Comment (304NotModified) [inline src/NLog/MappedDiagnosticsLogicalContext.cs]: why is this not needed for .NET 4.6?

File: src/NLog/MappedDiagnosticsLogicalContext.cs [hunk 2]
```diff
@@ -140,12 +151,23 @@ public static object GetObject(string item)
         /// <param name="item">Item name.</param>
         /// <param name="value">Item value.</param>
         /// <returns>>An <see cref="IDisposable"/> that can be used to remove the item from the current logical context.</returns>
-        public static IDisposable SetScoped(string item, string value)
+        public static IDisposable SetScoped<T>(string item, T value)
```
Comment (304NotModified) [inline src/NLog/MappedDiagnosticsLogicalContext.cs]: please move the generic below the non-generic, less potential conflict then

File: src/NLog/MappedDiagnosticsLogicalContext.cs [hunk 3]
```diff
@@ -175,7 +196,24 @@ public static void Set(string item, string value)
...
+        /// <param name="value">Item value.</param>
+        public static void Set<T>(string item, T value)
+        {
+#if NET4_6 || NETSTANDARD
             GetLogicalThreadDictionary(true)[item] = value;
+#else
+            if (Convert.GetTypeCode(value) != TypeCode.Object || typeof(T).IsValueType)
+                GetLogicalThreadDictionary(true)[item] = value;
+            else
+                GetLogicalThreadDictionary(true)[item] = new System.Runtime.Remoting.ObjectHandle(value);
```
Comment (304NotModified) [inline src/NLog/MappedDiagnosticsLogicalContext.cs]: please do  deduplicate the `GetLogicalThreadDictionary(true)[item]` set, e.g.

```c#
var val = GetSetValue(...)

GetLogicalThreadDictionary(true)[item] = val;
```

File: src/NLog/MappedDiagnosticsLogicalContext.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/MappedDiagnosticsLogicalContext.cs]: Because then I'm using AsyncLocal. Because NET Remoting is deprecated technology.

Comment (304NotModified) [reply src/NLog/MappedDiagnosticsLogicalContext.cs]: 👍

Comment (snakefoot) [reply src/NLog/MappedDiagnosticsLogicalContext.cs]: But this ofcourse require us to release a NET46 build of NLog to take effect :)

Comment (304NotModified) [reply src/NLog/MappedDiagnosticsLogicalContext.cs]: owjah yes indeed. 

More build time :'(

File: src/NLog/MappedDiagnosticsLogicalContext.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/MappedDiagnosticsLogicalContext.cs]: Fixed

File: src/NLog/MappedDiagnosticsLogicalContext.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/MappedDiagnosticsLogicalContext.cs]: Fixed

**Comment by snakefoot** [comment]
Discovered that ObjectHandle was a trap. Created #2910 to implement something better.

**Pull request merged**